File: packages/telegram-control/python/telegram_control/bot.py
"""Core bot implementation."""

import asyncio
import logging
from typing import Any, Awaitable, Callable, Dict, List, Optional, Pattern, Union

from telegram import Update
from telegram.ext import (
    Application,
    ApplicationBuilder,
    CallbackQueryHandler,
    CommandHandler,
    ContextTypes,
    MessageHandler,
    filters,
)

from .auth import AccessControl
from .types import CallbackContext, MessageContext

logger = logging.getLogger(__name__)

# Type aliases
HandlerFunc = Callable[[Any], Awaitable[None]]


class Bot:
    """Telegram Bot wrapper with trading-focused features."""
    
    def __init__(
        self,
        token: str,
        access_control: Optional[AccessControl] = None,
    ):
        self.token = token
        self.access_control = access_control or AccessControl()
        self._command_handlers: Dict[str, HandlerFunc] = {}
        self._callback_handlers: List[tuple] = []  # (pattern, handler)
        self._default_handler: Optional[HandlerFunc] = None
        self._application: Optional[Application] = None
    
    def with_whitelist(self, ids: List[int]) -> "Bot":
        """Restrict access to specific user IDs."""
        self.access_control.with_whitelist(ids)
        return self
    
    def with_admins(self, ids: List[int]) -> "Bot":
        """Set admin users."""
        self.access_control.with_admins(ids)
        return self
    
    def command(self, name: str):
        """Decorator for command handlers."""
        def decorator(func: HandlerFunc) -> HandlerFunc:
            self._command_handlers[f"/{name}"] = func
            self._command_handlers[name] = func
            return func
        return decorator
    
    def callback(self, pattern: Union[str, Pattern]):
        """Decorator for callback query handlers."""
        def decorator(func: HandlerFunc) -> HandlerFunc:
            self._callback_handlers.append((pattern, func))
            return func
        return decorator
    
    def default(self, func: HandlerFunc) -> HandlerFunc:
        """Set default handler for unmatched messages."""
        self._default_handler = func
        return func
    
    async def _handle_command(self, update: Update, context: ContextTypes.DEFAULT_TYPE):
        """Handle incoming commands."""
        if not update.message or not update.effective_user:
            return
        
        user_id = update.effective_user.id
        
        # Check authorization
        if not self.access_control.is_authorized(user_id):
            logger.warning(f"Unauthorized access attempt from user {user_id}")
            await update.message.reply_text("⛔ You are not authorized to use this bot.")
            return
        
        text = update.message.text or ""
        parts = text.split()
        cmd = parts[0] if parts else ""
        
        # Find handler
        handler = self._command_handlers.get(cmd)
        if handler:
            ctx = MessageContext(
                message=update.message,
                user=update.effective_user,
                chat_id=update.effective_chat.id if update.effective_chat else 0,
                text=text,
                bot=self,
            )
            try:
                await handler(ctx)
            except Exception as e:
                logger.exception("Command handler error")
                await update.message.reply_text(f"❌ Error: {e}")
        elif self._default_handler:
            ctx = MessageContext(
                message=update.message,
                user=update.effective_user,
                chat_id=update.effective_chat.id if update.effective_chat else 0,
                text=text,
                bot=self,
            )
            try:
                await self._default_handler(ctx)
            except Exception as e:
                logger.exception("Default handler error")
                await update.message.reply_text(f"❌ Error: {e}")
    
    async def _handle_callback(self, update: Update, context: ContextTypes.DEFAULT_TYPE):
        """Handle callback queries."""
        if not update.callback_query or not update.effective_user:
            return
        
        query = update.callback_query
        data = query.data or ""
        
        # Find matching handler
        for pattern, handler in self._callback_handlers:
            import re
            if isinstance(pattern, str):
                if data.startswith(pattern):
                    ctx = CallbackContext(
                        query=query,
                        user=update.effective_user,
                        chat_id=update.effective_chat.id if update.effective_chat else 0,
                        data=data,
                        bot=self,
                    )
                    try:
                        await handler(ctx)
                    except Exception as e:
                        logger.exception("Callback handler error")
                        await query.answer(f"Error: {e}")
                    return
            elif isinstance(pattern, Pattern):
                if pattern.match(data):
                    ctx = CallbackContext(
                        query=query,
                        user=update.effective_user,
                        chat_id=update.effective_chat.id if update.effective_chat else 0,
                        data=data,
                        bot=self,
                    )
                    try:
                        await handler(ctx)
                    except Exception as e:
                        logger.exception("Callback handler error")
                        await query.answer(f"Error: {e}")
                    return
    
    def build(self) -> Application:
        """Build the Application instance."""
        app = ApplicationBuilder().token(self.token).build()
        
        # Add command handler
        app.add_handler(
            MessageHandler(filters.COMMAND | filters.TEXT, self._handle_command)
        )
        
        # Add callback handler
        if self._callback_handlers:
            app.add_handler(CallbackQueryHandler(self._handle_callback))
        
        self._application = app
        return app
    
    def run(self, poll_interval: float = 1.0):
        """Run the bot (blocking)."""
        app = self.build()
        logger.info("Starting Telegram bot...")
        app.run_polling(poll_interval=poll_interval)
    
    async def run_async(self, poll_interval: float = 1.0):
        """Run the bot (async)."""
        app = self.build()
        logger.info("Starting Telegram bot...")
        await app.initialize()
        await app.start()
        await app.updater.start_polling(poll_interval=poll_interval)
    
    async def send_message(self, chat_id: int, text: str, **kwargs):
        """Send a message to a chat."""
        if self._application:
            await self._application.bot.send_message(chat_id=chat_id, text=text, **kwargs)
    
    async def send_alert(self, chat_id: int, text: str, **kwargs):
        """Send an alert (formatted message)."""
        await self.send_message(chat_id, text, parse_mode="Markdown", **kwargs)

File: packages/telegram-control/rust/src/bot.rs
use std::collections::HashMap;
use std::future::Future;
use std::pin::Pin;
use std::sync::Arc;

use teloxide::dispatching::{Dispatcher, UpdateFilterExt, HandlerExt};
use teloxide::prelude::*;
use teloxide::types::Update;
use teloxide::utils::command::BotCommands;
use tracing::{info, warn, error};

use crate::auth::AccessControl;
use crate::error::{Error, Result};
use crate::types::{CallbackContext, Context, MessageContext};

/// Handler function type
pub type HandlerFn = Arc<
    dyn Fn(Context) -> Pin<Box<dyn Future<Output = Result<()>> + Send>> + Send + Sync,
>;

/// Telegram Bot wrapper with trading-focused features
pub struct Bot {
    bot: teloxide::Bot,
    access_control: AccessControl,
    command_handlers: HashMap<String, HandlerFn>,
    callback_handlers: Vec<(String, HandlerFn)>, // pattern, handler
    default_handler: Option<HandlerFn>,
}

impl Bot {
    /// Create a new bot instance
    pub fn new(token: impl Into<String>) -> BotBuilder {
        BotBuilder::new(token)
    }
    
    /// Register a command handler
    pub fn on_command<F, Fut>(mut self, command: impl Into<String>, handler: F) -> Self
    where
        F: Fn(Context) -> Fut + Send + Sync + 'static,
        Fut: Future<Output = Result<()>> + Send + 'static,
    {
        let handler: HandlerFn = Arc::new(move |ctx| Box::pin(handler(ctx)));
        self.command_handlers.insert(command.into(), handler);
        self
    }
    
    /// Register a callback query handler with pattern matching
    pub fn on_callback<F, Fut>(mut self, pattern: impl Into<String>, handler: F) -> Self
    where
        F: Fn(Context) -> Fut + Send + Sync + 'static,
        Fut: Future<Output = Result<()>> + Send + 'static,
    {
        let handler: HandlerFn = Arc::new(move |ctx| Box::pin(handler(ctx)));
        self.callback_handlers.push((pattern.into(), handler));
        self
    }
    
    /// Set default handler for unmatched messages
    pub fn on_default<F, Fut>(mut self, handler: F) -> Self
    where
        F: Fn(Context) -> Fut + Send + Sync + 'static,
        Fut: Future<Output = Result<()>> + Send + 'static,
    {
        self.default_handler = Some(Arc::new(move |ctx| Box::pin(handler(ctx))));
        self
    }
    
    /// Run the bot (blocking)
    pub async fn run(self) -> Result<()> {
        info!("Starting Telegram bot...");
        
        let bot = self.bot.clone();
        let access_control = self.access_control.clone();
        let command_handlers = Arc::new(self.command_handlers);
        let callback_handlers = Arc::new(self.callback_handlers);
        let default_handler = self.default_handler;
        
        let handler = dptree::entry()
            .branch(
                Update::filter_message().endpoint(
                    move |bot: teloxide::Bot, msg: Message| {
                        let access_control = access_control.clone();
                        let handlers = command_handlers.clone();
                        let default = default_handler.clone();
                        
                        async move {
                            if let Some(user) = msg.from() {
                                let user_id = user.id.0 as i64;
                                
                                // Check authorization
                                if let Err(e) = access_control.authorize(user_id) {
                                    warn!("Unauthorized access attempt from user {}", user_id);
                                    let _ = bot.send_message(
                                        msg.chat.id,
                                        "⛔ You are not authorized to use this bot."
                                    ).await;
                                    return Ok(());
                                }
                                
                                let ctx = MessageContext {
                                    message: msg.clone(),
                                    user: user.clone(),
                                    chat_id: msg.chat.id.0,
                                    text: msg.text().map(|s| s.to_string()),
                                    bot: bot.clone(),
                                };
                                
                                // Try to match command
                                if let Some(text) = msg.text() {
                                    let parts: Vec<&str> = text.split_whitespace().collect();
                                    if let Some(cmd) = parts.first() {
                                        if let Some(handler) = handlers.get(*cmd) {
                                            if let Err(e) = handler(Context::Message(ctx)).await {
                                                error!("Handler error: {}", e);
                                                let _ = bot.send_message(
                                                    msg.chat.id,
                                                    format!("❌ Error: {}", e)
                                                ).await;
                                            }
                                            return Ok(());
                                        }
                                    }
                                }
                                
                                // Default handler
                                if let Some(handler) = default {
                                    if let Err(e) = handler(Context::Message(ctx)).await {
                                        error!("Default handler error: {}", e);
                                    }
                                }
                            }
                            Ok(())
                        }
                    }
                )
            )
            .branch(
                Update::filter_callback_query().endpoint(
                    move |bot: teloxide::Bot, q: CallbackQuery| {
                        let handlers = callback_handlers.clone();
                        
                        async move {
                            if let (Some(data), Some(user)) = (q.data.clone(), q.from) {
                                let ctx = CallbackContext {
                                    query: q.clone(),
                                    user: user.clone(),
                                    chat_id: q.message.as_ref()
                                        .map(|m| m.chat.id.0)
                                        .unwrap_or(user.id.0 as i64),
                                    data: data.clone(),
                                    bot: bot.clone(),
                                };
                                
                                // Find matching handler
                                for (pattern, handler) in handlers.iter() {
                                    if data.starts_with(pattern) {
                                        if let Err(e) = handler(Context::Callback(ctx)).await {
                                            error!("Callback handler error: {}", e);
                                        }
                                        return Ok(());
                                    }
                                }
                            }
                            Ok(())
                        }
                    }
                )
            );
        
        Dispatcher::builder(bot, handler)
            .enable_ctrlc_handler()
            .build()
            .dispatch()
            .await;
        
        Ok(())
    }
}

/// Builder for Bot configuration
pub struct BotBuilder {
    token: String,
    access_control: AccessControl,
}

impl BotBuilder {
    pub fn new(token: impl Into<String>) -> Self {
        Self {
            token: token.into(),
            access_control: AccessControl::new(),
        }
    }
    
    pub fn with_whitelist(mut self, ids: Vec<i64>) -> Self {
        self.access_control = self.access_control.with_whitelist(ids);
        self
    }
    
    pub fn with_admins(mut self, ids: Vec<i64>) -> Self {
        self.access_control = self.access_control.with_admins(ids);
        self
    }
    
    pub fn build(self) -> Bot {
        Bot {
            bot: teloxide::Bot::new(self.token),
            access_control: self.access_control,
            command_handlers: HashMap::new(),
            callback_handlers: Vec::new(),
            default_handler: None,
        }
    }
}

impl From<BotBuilder> for Bot {
    fn from(builder: BotBuilder) -> Self {
        builder.build()
    }
}
